src/test/scala/org/apache/spark/sql/delta/MergeIntoSuiteBase.scala

```diff
@@ -1718,4 +1718,46 @@ abstract class MergeIntoSuiteBase
       }
     }
   }
+
+  protected def testMatchedOnlyOptimization(
+                                             name: String)(
+                                             source: Seq[(Int, Int)],
+                                             target: Seq[(Int, Int)],
+                                             mergeOn: String,
+                                             mergeClauses: MergeClause*) (
+                                             result: Seq[(Int, Int)]): Unit = {
+    Seq(true, false).foreach { matchedOnlyEnabled =>
+      withSQLConf(DeltaSQLConf.MERGE_MATCHED_ONLY_ENABLED.key -> matchedOnlyEnabled.toString) {
+        testExtendedMerge(
+          s"matched only clause enabled = ${matchedOnlyEnabled} - $name") (
+          source,
+          target,
+          mergeOn,
+          mergeClauses: _*) (
+          result
+        )
+      }
+    }
+  }
+
+  testMatchedOnlyOptimization("with update") (
+    source = Seq((1, 100), (3, 300), (5, 500)),
+    target = Seq((1, 10), (2, 20), (3, 30)),
+    mergeOn = "s.key = t.key",
+    update("t.key = s.key, t.value = s.value")) (
+    result = Seq(
+      (1, 100), // updated
+      (2, 20), // existed previously
+      (3, 300) // updated
+    )
+  )
+  testMatchedOnlyOptimization("with delete") (
+    source = Seq((1, 100), (3, 300), (5, 500)),
+    target = Seq((1, 10), (2, 20), (3, 30)),
+    mergeOn = "s.key = t.key",
+    delete()) (
+    result = Seq(
+      (2, 20) // existed previously
+    )
+  )
```

Comment:
need additional tests for black-box testing of the matched only queries 
- with update and delete clause conditions
- null tests
(similar to the existing tests for insert-only-merge)